src/main/java/marquez/core/services/JobService.java

```diff
@@ -0,0 +1,224 @@
+    try {
+      return Optional.ofNullable(jobVersionDAO.findLatest(namespace, jobName));
+    } catch (UnableToExecuteStatementException e) {
+      String err = "error fetching latest version of job";
+      log.error(err, e);
+      throw new UnexpectedException();
+    }
+  }
+
+  public JobRun updateJobRunState(UUID jobRunID, JobRunState.State state)
+      throws UnexpectedException {
+    try {
+      this.jobRunDAO.updateState(jobRunID, JobRunState.State.toInt(state));
+      return this.jobRunDAO.findJobRunById(jobRunID);
+    } catch (UnableToExecuteStatementException e) {
+      String err = "error updating job run state";
+      log.error(err, e);
+      throw new UnexpectedException();
+    }
+  }
+
+  public Optional<JobRun> getJobRun(UUID jobRunID) throws UnexpectedException {
+    try {
+      return Optional.ofNullable(this.jobRunDAO.findJobRunById(jobRunID));
+    } catch (UnableToExecuteStatementException e) {
+      String err = "error fetching job run";
+      log.error(err, e);
+      throw new UnexpectedException();
+    }
+  }
+
+  public JobRun createJobRun(
+      String namespaceName,
+      String jobName,
+      String runArgsJson,
+      Timestamp nominalStartTime,
+      Timestamp nominalEndTime)
+      throws UnexpectedException {
+    try {
+      String runArgsDigest = null;
+      RunArgs runArgs = null;
+      if (null == jobDAO.findByName(namespaceName, jobName)) {
+        String err =
+            String.format(
+                "unable to find job <ns='%s', job name='%s'> to create job run",
+                namespaceName, jobName);
+        log.error(err);
+        throw new UnexpectedException();
+      }
+      Optional<JobVersion> latestJobVersion = getLatestVersionOfJob(namespaceName, jobName);
+      if (!latestJobVersion.isPresent()) {
+        String err =
+            String.format(
+                "unable to find latest job version for <ns='%s', job name='%s'> to create job run",
+                namespaceName, jobName);
+        log.error(err);
+        throw new UnexpectedException();
+      }
+      if (runArgsJson != null) {
+        runArgsDigest = computeRunArgsDigest(runArgsJson);
+        runArgs = new RunArgs(runArgsDigest, runArgsJson, null);
+      }
+      JobRun jobRun =
+          new JobRun(
+              UUID.randomUUID(),
+              JobRunState.State.toInt(JobRunState.State.NEW),
+              latestJobVersion.get().getGuid(),
+              runArgsDigest,
+              runArgsJson,
+              nominalStartTime,
+              nominalEndTime,
+              null);
+      if (runArgsJson == null || runArgsDAO.digestExists(runArgsDigest)) {
+        jobRunDAO.insert(jobRun);
+      } else {
+        jobRunDAO.insertJobRunAndArgs(jobRun, runArgs);
+      }
+      return jobRun;
+    } catch (UnableToExecuteStatementException | NoSuchAlgorithmException e) {
+      String err = "error creating job run";
+      log.error(err, e);
+      throw new UnexpectedException();
```

Comment:
Can we factor this out since it's used in a few places?

Reply:
while the pattern is repeated, each one has a potentially different set of exceptions to catch and throw. at the moment they're the all the same, but they will vary independently. Refactoring it is not likely to cut down any code, only make it less idiomatic

Reply:
Ok cool :)